reco_utils/recommender/sar/sar_singlenode.py

```diff
@@ -208,97 +200,116 @@ def fit(self, df):
+            )
         else:
             raise ValueError(
                 "Unknown similarity type: {0}".format(self.similarity_type)
             )
 
-        # Calculate raw scores with a matrix multiplication
-        logger.info("Calculating recommendation scores")
-        self.scores = self.user_affinity.dot(self.item_similarity)
-
-        # Remove items in the train set so recommended items are always novel
-        if self.remove_seen:
-            logger.info("Removing seen items")
-            self.scores[seen_items[:, 0], seen_items[:, 1]] = -np.inf
+        # free up some space
+        del item_cooccurrence
 
         logger.info("Done training")
 
-    def recommend_k_items(self, test, top_k=10, sort_top_k=False):
-        """Recommend top K items for all users which are in the test set
+    def score(self, test, remove_seen=False):
+        """Score all items for test users
 
         Args:
             test (pd.DataFrame): user to test
-            top_k (int): number of top items to recommend
-            sort_top_k (bool): flag to sort top k results
+            remove_seen (bool): flag to remove items seen in training from recommendation
         Returns:
-            pd.DataFrame: top k recommendation items for each user
+            np.ndarray
         """
 
         # get user / item indices from test set
         user_ids = test[self.col_user].drop_duplicates().map(self.user2index).values
         if any(np.isnan(user_ids)):
             raise ValueError("SAR cannot score users that are not in the training set")
 
-        # extract only the scores for the test users
-        test_scores = self.scores[user_ids, :]
+        # calculate raw scores with a matrix multiplication
+        logger.info("Calculating recommendation scores")
+        # TODO: only compute scores for users in test
+        test_scores = self.user_affinity.dot(self.item_similarity)
+
+        # remove items in the train set so recommended items are always novel
+        if remove_seen:
+            logger.info("Removing seen items")
+            test_scores[self.seen_items[:, 0], self.seen_items[:, 1]] = -np.inf
+
+        test_scores = test_scores[user_ids, :]
 
         # ensure we're working with a dense matrix
         if isinstance(test_scores, sparse.spmatrix):
             test_scores = test_scores.todense()
 
+        return test_scores
+
+    def recommend_k_items(self, test, top_k=10, sort_top_k=False, remove_seen=False):
+        """Recommend top K items for all users which are in the test set
+
+        Args:
+            test (pd.DataFrame): user to test
+            top_k (int): number of top items to recommend
+            sort_top_k (bool): flag to sort top k results
+            remove_seen (bool): flag to remove items seen in training from recommendation
+        Returns:
+            pd.DataFrame: top k recommendation items for each user
+        """
+
+        k = min(top_k, self.n_items)
```

Comment:
Maybe here we should have a warning in case num_items < k, so the user is aware that the k is modified